### PR TITLE
fix(remix): RequestInit.duplex now is required if body is provided in undici

### PR DESCRIPTION
**Pull request by octoper**
## Description

RequestInit.duplex now is required if body is provided:
- https://github.com/nodejs/node/issues/46221
- https://github.com/whatwg/fetch/pull/1457

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/backend | 1.2.2-snapshot.v72984d5 |
| @clerk/chrome-extension | 1.0.17-snapshot.v72984d5 |
| @clerk/clerk-js | 5.6.0-snapshot.v72984d5 |
| @clerk/elements | 0.6.0-snapshot.v72984d5 |
| @clerk/clerk-expo | 1.1.9-snapshot.v72984d5 |
| @clerk/express | 0.0.11-snapshot.v72984d5 |
| @clerk/fastify | 1.0.13-snapshot.v72984d5 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/localizations | 2.4.4-snapshot.v72984d5 |
| @clerk/nextjs | 5.1.4-snapshot.v72984d5 |
| @clerk/clerk-react | 5.2.3-snapshot.v72984d5 |
| @clerk/remix | 4.0.14-snapshot.v72984d5 |
| @clerk/clerk-sdk-node | 5.0.10-snapshot.v72984d5 |
| @clerk/shared | 2.2.2-snapshot.v72984d5 |
| @clerk/testing | 1.1.6-snapshot.v72984d5 |
| @clerk/themes | 2.1.9-snapshot.v72984d5 |
| @clerk/types | 4.6.0-snapshot.v72984d5 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/backend`
```sh
npm i @clerk/backend@1.2.2-snapshot.v72984d5 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.0.17-snapshot.v72984d5 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.6.0-snapshot.v72984d5 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.6.0-snapshot.v72984d5 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.1.9-snapshot.v72984d5 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.11-snapshot.v72984d5 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.13-snapshot.v72984d5 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@2.4.4-snapshot.v72984d5 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.1.4-snapshot.v72984d5 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.2.3-snapshot.v72984d5 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.0.14-snapshot.v72984d5 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.10-snapshot.v72984d5 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.2.2-snapshot.v72984d5 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.6-snapshot.v72984d5 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.9-snapshot.v72984d5 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.6.0-snapshot.v72984d5 --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/backend | 1.2.2-snapshot.vef44e76 |
| @clerk/chrome-extension | 1.0.17-snapshot.vef44e76 |
| @clerk/clerk-js | 5.6.0-snapshot.vef44e76 |
| @clerk/elements | 0.6.0-snapshot.vef44e76 |
| @clerk/clerk-expo | 1.1.9-snapshot.vef44e76 |
| @clerk/express | 0.0.11-snapshot.vef44e76 |
| @clerk/fastify | 1.0.13-snapshot.vef44e76 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/localizations | 2.4.4-snapshot.vef44e76 |
| @clerk/nextjs | 5.1.4-snapshot.vef44e76 |
| @clerk/clerk-react | 5.2.3-snapshot.vef44e76 |
| @clerk/remix | 4.0.14-snapshot.vef44e76 |
| @clerk/clerk-sdk-node | 5.0.10-snapshot.vef44e76 |
| @clerk/shared | 2.2.2-snapshot.vef44e76 |
| @clerk/testing | 1.1.6-snapshot.vef44e76 |
| @clerk/themes | 2.1.9-snapshot.vef44e76 |
| @clerk/types | 4.6.0-snapshot.vef44e76 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/backend`
```sh
npm i @clerk/backend@1.2.2-snapshot.vef44e76 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.0.17-snapshot.vef44e76 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.6.0-snapshot.vef44e76 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.6.0-snapshot.vef44e76 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.1.9-snapshot.vef44e76 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.11-snapshot.vef44e76 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.13-snapshot.vef44e76 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@2.4.4-snapshot.vef44e76 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.1.4-snapshot.vef44e76 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.2.3-snapshot.vef44e76 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.0.14-snapshot.vef44e76 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.10-snapshot.vef44e76 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.2.2-snapshot.vef44e76 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.6-snapshot.vef44e76 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.9-snapshot.vef44e76 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.6.0-snapshot.vef44e76 --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/backend | 1.2.2-snapshot.vff20ad3 |
| @clerk/chrome-extension | 1.0.17-snapshot.vff20ad3 |
| @clerk/clerk-js | 5.6.0-snapshot.vff20ad3 |
| @clerk/elements | 0.6.0-snapshot.vff20ad3 |
| @clerk/clerk-expo | 1.1.9-snapshot.vff20ad3 |
| @clerk/express | 0.0.11-snapshot.vff20ad3 |
| @clerk/fastify | 1.0.13-snapshot.vff20ad3 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/localizations | 2.4.4-snapshot.vff20ad3 |
| @clerk/nextjs | 5.1.4-snapshot.vff20ad3 |
| @clerk/clerk-react | 5.2.3-snapshot.vff20ad3 |
| @clerk/remix | 4.0.14-snapshot.vff20ad3 |
| @clerk/clerk-sdk-node | 5.0.10-snapshot.vff20ad3 |
| @clerk/shared | 2.2.2-snapshot.vff20ad3 |
| @clerk/testing | 1.1.6-snapshot.vff20ad3 |
| @clerk/themes | 2.1.9-snapshot.vff20ad3 |
| @clerk/types | 4.6.0-snapshot.vff20ad3 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/backend`
```sh
npm i @clerk/backend@1.2.2-snapshot.vff20ad3 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.0.17-snapshot.vff20ad3 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.6.0-snapshot.vff20ad3 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.6.0-snapshot.vff20ad3 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.1.9-snapshot.vff20ad3 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.11-snapshot.vff20ad3 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.13-snapshot.vff20ad3 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@2.4.4-snapshot.vff20ad3 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.1.4-snapshot.vff20ad3 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.2.3-snapshot.vff20ad3 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.0.14-snapshot.vff20ad3 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.10-snapshot.vff20ad3 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.2.2-snapshot.vff20ad3 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.6-snapshot.vff20ad3 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.9-snapshot.vff20ad3 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.6.0-snapshot.vff20ad3 --save-exact
```

File: .changeset/fifty-parents-juggle.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@clerk/backend': patch
```
Comment (anagstef) [inline .changeset/fifty-parents-juggle.md:2]:
```suggestion
'@clerk/remix': patch
```

Comment (octoper) [reply .changeset/fifty-parents-juggle.md]: Nice catch! 💯

File: packages/remix/src/ssr/utils.ts [hunk 2]
```diff
@@ -116,3 +116,18 @@ export function getResponseClerkState(requestState: RequestStateWithRedirectUrls
...
 };
+
+/**
+ * Patches request to avoid duplex issues with unidici
+ * For more information, see:
+ * https://github.com/nodejs/node/issues/46221
+ * https://github.com/whatwg/fetch/pull/1457
+ * @internal
+ */
+export const patchRequest = (request: Request) => {
```
Comment (anagstef) [inline packages/remix/src/ssr/utils.ts:127]: @octoper 💡  maybe rename this one to reflect the specific patch? For example `patchRequestDuplex`, or something similar? 🤔

**Review by anagstef — APPROVED** [review]
💯

File: packages/remix/src/ssr/utils.ts (hunk 2 above)
Comment (octoper) [reply packages/remix/src/ssr/utils.ts:127]: Named it `patchRequest` in case we would need to patch other things in the future 😅

**Comment by ariofrio** [comment]
I'm getting the following error:

```
Error: cannot clone body after it is used
    at clone (.../build/server.js:36692:15)
    at new Request2 (.../build/server.js:37302:99)
    at new NodeRequest (.../build/server.js:37754:9)
    at NodeRequest.clone (.../build/server.js:37760:16)
    at patchRequest (.../node_modules/@clerk/remix/src/ssr/utils.ts:128:33)
    at loadOptions (.../node_modules/@clerk/remix/src/ssr/loadOptions.ts:15:43)
    at Object.getAuth2 (.../node_modules/@clerk/remix/src/ssr/getAuth.ts:15:25)
```

The problem is that my code looks like this:

```ts
export async function action(args: ActionArgs) {
  const formData = await args.request.formData();
  const userId = await getAuth(args);
  ...
}
```

In other words, the request body is read before calling `clerk.getAuth`. Since Clerk attempts to clone the request, this causes an error. Is it necessary to clone the request to work around this duplex issue?

The workaround is to always call getAuth first:


```ts
export async function action(args: ActionArgs) {
  const userId = await getAuth(args);
  const formData = await args.request.formData();
  ...
}
```

But this is error-prone and easy to forget to do or accidentally change in a large project. At the minimum, this should be documented in the Clerk Remix guide. Ideally, this limitation would be removed.

**Pull request merged**